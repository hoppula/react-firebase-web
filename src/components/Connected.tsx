import * as PropTypes from "prop-types"
import * as React from "react"
import { FirebaseContext } from "../types"

export interface ConnectedProps {
  readonly children: (connected: boolean) => React.ReactElement<any>
}

export interface ConnectedState {
  connected: boolean
}

export class Connected extends React.Component<ConnectedProps, ConnectedState> {
  static propTypes = {
    children: PropTypes.func.isRequired
  }

  static contextTypes = {
    firebase: PropTypes.object.isRequired
  }

  context: FirebaseContext
  state: ConnectedState = { connected: false }

  componentDidMount() {
    this.context.firebase
      .database()
      .ref(".info/connected")
      .on("value", snap => {
        this.setState({ connected: snap.val() === true })
      })
  }

  render() {
    return React.Children.only(this.props.children(this.state.connected))
  }
}

export default Connected
