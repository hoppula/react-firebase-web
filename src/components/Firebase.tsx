import * as firebase from "firebase"
import * as PropTypes from "prop-types"
import * as React from "react"

export interface FirebaseProps {
  readonly apiKey: string
  readonly projectId: string
  readonly databaseURL?: string
}

export class Firebase extends React.Component<FirebaseProps, {}> {
  static propTypes = {
    apiKey: PropTypes.string.isRequired,
    databaseURL: PropTypes.string,
    projectId: PropTypes.string.isRequired
  }

  static childContextTypes = {
    firebase: PropTypes.object
  }

  firebase: firebase.app.App

  constructor(props: FirebaseProps) {
    super(props)
    const { apiKey, projectId, databaseURL } = props
    this.firebase = firebase.initializeApp({
      apiKey,
      authDomain: `${projectId}.firebaseapp.com`,
      databaseURL: databaseURL
        ? databaseURL
        : `https://${projectId}.firebaseio.com`,
      storageBucket: `${projectId}.appspot.com`
    })
  }

  getChildContext() {
    return {
      firebase: this.firebase
    }
  }

  render() {
    return React.Children.only(this.props.children)
  }
}

export default Firebase
