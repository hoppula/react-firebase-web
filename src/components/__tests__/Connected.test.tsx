import React from "react"
import renderer from "react-test-renderer"
import { Firebase } from "../Firebase"
import Connected from "../Connected"

import { initializeMockSDK } from "./setup"
const mocksdk = initializeMockSDK()

jest.mock("firebase/app", () => {
  return {
    initializeApp: (...args: any[]) => mocksdk.initializeApp(...args)
  }
})

describe("Connected", () => {
  test("render prop value is set to true when online", () => {
    let component = renderer.create(
      <Firebase apiKey="firebase-test" projectId="firebase-test">
        <Connected
          children={connected => {
            if (connected) {
              return <div>Online :)</div>
            } else {
              return <div>Offline :(</div>
            }
          }}
        />
      </Firebase>
    )

    // offline
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    const ref = component
      .getInstance()
      .firebase.database()
      .ref(".info/connected")
    ref.set(true)
    ref.flush()

    // online
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
