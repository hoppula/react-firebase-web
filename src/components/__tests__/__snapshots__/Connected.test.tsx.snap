// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected render prop value is set to true when online 1`] = `
<div>
  Offline :(
</div>
`;

exports[`Connected render prop value is set to true when online 2`] = `
<div>
  Online :)
</div>
`;
